DietPi-Software | "/tmp/dietpi-downloads" and unified download+install function

MichaIng opened the pull request:
**Status**: Ready for testing
- [x] Test affected software installations
  - x86_64 VirtualBox
  - 🈯️ NoMachine (+ Java, Xserver, LXDE, ALSA)
  - 🈯️ ownCloud (+ MariaDB, Lighttpd, PHP, Redis)
  - 🈯️ MPD, phpBB
  - 🈯️ OpenBazaar: Puhh, building within ramdisk doesn't speed up at all on VM. CPU is cap here... running already for 20 Minutes or what... Done finally, and working
  - 🈯️ `p7zip-full` on Jessie does not contain `7zr` (as on Stretch+), pre-install `p7zip` or use less lightweight `7za` or `7z` (generally or on Jessie only?)? **€: Fixed with https://github.com/Fourdee/DietPi/pull/1906/commits/de0b0ae56fecbbc09590808242aef7408413a074, using 7za for now**
  - 🈯️ wiringPi needs to stay within `$HOME`? **€: Move it to $HOME and leave it there, as it seems expected location: https://github.com/Fourdee/DietPi/pull/1906/commits/de0b0ae56fecbbc09590808242aef7408413a074**
  - 🈯️ All affected titles, that are available for x86_64, tested on Jessie and Stretch and some on Buster as well.
- [x] Verify `tar xf` extracts `tar.gz` and `tar.bz[2]` without further argument and unified `.tar` file ending
  - Did some tests already, but should further be tested with all kind of tarballs.
- [x] Phase 2: Implement `Download_Install` into further installations, where applicable
- [x] Phase 3: Implement `Download_Install` into further installations, where applicable
- [x] Phase 4: Implement `Download_Install` into further installations, where applicable
<s>- [ ] Optional: Further merge similar `Download_Install` archive processing into single `if` statement, using extract command variable</s>
- [x] Optional: Make `/tmp/dietpi-software` the default working directory `$FP_WORK_DIR` for `dietpi-software` instead of `$HOME`
- [x] Optional: Add `$URL_LIST` to allow checking all needed URLs (dependencies + later Download_Install executions) before starting any download
<s>- [ ] Optional: Add `.sh`, `.jar` and `git` clone handling</s>

**Reference**: https://github.com/Fourdee/DietPi/issues/1801
- Actually I just wanted to start implementing `/tmp/dietpi-downloads` a bid, but I couldn't hold myself to realize an idea, that haunted my mind for a while 😄.

**Commit list/description**:
+ DietPi-Software | New Download_Install function that downloads packages/archives to /tmp RAMdisk, installs/extracts them based on type and cleans up afterwards
+ DietPi-Software | "Download_Install()" installs deb package dependencies based on $DEPS_LIST and in case starts download as G_THREAD_START
+ DietPi-Software | Install all needed APT packages for a software title with one G_AGI call, to speed up installation (APT database read/write) and reduce redundant output
+ DietPi-Software | Minor code improvements
+ DietPi-Software | Download_Install: Phase 2 integration
+ DietPi-Software | Download_Install: Add optional "$no_check_url" flag
+ DietPi-Software | Download_Install: On dpkg error, try to fix with G_AGF, which includes automated dependency installation
+ DietPi-Software | Download_Install: Automatically pre-create archive target folder
+ DietPi-Software | MPD: Add/Fix Buster support
+ DietPi-Software | Fix missing $FP_DL_TMP additions
+ DietPi-Software | Download_Install: Handle $DEPS_LIST for all download types
+ DietPi-Software | CertBot: Always install "certbot" package, to prevent auto removal, if webserver is switched
+ DietPi-Software | Download_Install: Use 7za instead of 7zr for now, as 7zr is not available on Jessie by default (p7zip-full does not depend on p7zip on Jessie)
+ DietPi-Software | Download_Install implementation Phase 3
+ DietPi-Software | Download_Install: Implement $URL_LIST to allow pre-checking all URLs each software title, before starting any download
+ DietPi-Software | Switch to $FP_WORK_DIR=/tmp/dietpi-software for all file creations and downloads of this script
+ DietPi-Software | Download_Install: Implementation Phase 4
+ DietPi-Software | Sickrage: Fix installation failing due to not yet chowned directory
+ DietPi-Software | Remove $FP_WORK_DIR (/tmp/dietpi-software) completely on exit

MichaIng committed:
v6.12

+ DietPi-Software | New Download_Install function that downloads packages/archives to /tmp RAMdisk, installs/extracts them based on type and cleans up afterwards
+ DietPi-Software | "Download_Install()" installs deb package dependencies based on $DEPS_LIST and in case starts download as G_THREAD_START
+ DietPi-Software | Install all needed APT packages for a software title with one G_AGI call, to speed up installation (APT database read/write) and reduce redundant output
+ DietPi-Software | Minor code improvements

MichaIng committed:
v6.12

+ DietPi-Software | Download_Install: Whoopsie + minor wording

MichaIng committed:
v6.12

+ DietPi-Software | Download_Install: Phase 2 integration
+ DietPi-Software | Download_Install: Add optional "$no_check_url" flag
+ DietPi-Software | Download_Install: On dpkg error, try to fix with G_AGF, which includes automated dependency installation
+ DietPi-Software | Download_Install: Automatically pre-create archive target folder

MichaIng committed:
v6.12

DietPi-Software | Syntax

MichaIng committed:
v6.12

+ DietPi-Software | MPD: Add/Fix Buster support

MichaIng commented:
@Fourdee 
Running OpenBazaar installation+compiling in RAM takes 650M usage peek, maybe more, didn't watch all the time.
With 2G due to swap file it should be no issue, but on <1G RAM systems loads to swap usage, which itself is already not good on such huge long time consuming compiling...

I have really not much knowledge about this, but it should be possible to first clone, build and install libzmq, then clean up /tmp/dietpi-downloads, then the same with pynacl and at last after again cleaning OpenBazaar, right? This would at least reduce data in /tmp ramdisk during the process.

Fourdee commented:
@MichaIng 

> but it should be possible to first clone, build and install libzmq, then clean up /tmp/dietpi-downloads, then the same with pynacl and at last after again cleaning OpenBazaar, right? This would at least reduce data in /tmp ramdisk during the process.

Yep, it can, once its compiled and installed (```make install ```):
```
			# - compile
			cd "$HOME"/libzmq
			./autogen.sh && ./configure && make -j $G_HW_CPU_CORES
			make check && make install && ldconfig
```
```"$HOME"/libzmq``` in this case, could be removed.

That is, unless ```pynacl``` requires the source code contained in ```"$HOME"/libzmq```, which should not be the case. 
As ```pynacl``` is Python and wont be able to read/share C++ source code of ```"$HOME"/libzmq```, as far as i'am aware.

Regardless, usually, shared libraries are installed on the system, which can be shared across the required compiling items, eg ```libboost-dev``` (```libboost-dev``` is not required, unless, your a lazy programmer). However, Python is a little different, does its own thing (~which I know nothing of~).

```
python-dev - header files and a static library for Python (default)
```
The rest of python libs come from ```pip``` https://stackoverflow.com/questions/21222114/how-do-i-install-python-libraries

Happy days 😃 😉 🍷

MichaIng commented:
@Fourdee 
Just another question:
- For wiringPi we download the src to $HOME, build it there, but don't move it to somewhere else or remove the folder.
- Does this need to stay within $HOME dir, as I guess ./build does not include any system installation, or does it?

And:
- Do you agree to use /tmp/dietpi-downloads or then maybe /tmp/dietpi-software as default working directory for the script? There are so much single downloads where cd here cd back would be needed otherwise.
- I assume that using $HOME in all our scripts is due to assuring write permissions for whatever user starts the script, having a fixed directory to always return to and move, where ever the script is started from and not have downloads/tmp files created across the system, if cd is used within sub function e.g. But generally, when script requires root user anyway, it should not matter which directory we use for this, and in this case /tmp ramdisk has discussed benefits.

MichaIng committed:
v6.12

+ DietPi-Software | Fix missing $FP_DL_TMP additions
+ DietPi-Software | Download_Install: Handle $DEPS_LIST for all download types
+ DietPi-Software | CertBot: Always install "certbot" package, to prevent auto removal, if webserver is switched
+ DietPi-Software | Minor coding

MichaIng committed:
Merge branch 'testing' into dietpi-software

MichaIng committed:
v6.12

+ DietPi-Software | Minor

MichaIng committed:
v6.12

+ DietPi-Software | Download_Install implementation Phase 3

Fourdee commented:
@MichaIng 

> I assume that using $HOME in all our scripts is due to assuring write permissions for whatever user starts the script, having a fixed directory to always return to and move, where ever the script is started from and not have downloads/tmp files created across the system

Yep, ensures we know we are in a single/same directory at all times.

> For wiringPi we download the src to $HOME, build it there, but don't move it to somewhere else or remove the folder.

I believe its the default known location for where it "should be" once installed.

> Do you agree to use /tmp/dietpi-downloads or then maybe /tmp/dietpi-software as default working directory for the script? There are so much single downloads where cd here cd back would be needed otherwise.

Yep, by all means go ahead. Either is fine, for simplicity i'd probably go with a shared pool ```/tmp/dietpi-downloads``` that we can use across scripts. Could even set a subdir if required:
```
/tmp/dietpi-downloads/dietpi-software
```

MichaIng commented:
@Fourdee 
> I believe its the default known location for where it "should be" once installed.

👍 Good, it is moved there now. Any other changes, e.g. run user and therefore dependent different $HOME dir (?) should be not part of this PR. It is large enough 😄.

> i'd probably go with a shared pool /tmp/dietpi-downloads that we can use across scripts. Could even set a subdir if required

I thought little bit about general script working dir directories:
- I think safest solution is to let every script use it's own /tmp/dietpi-* directory, so within we exactly know what is expected inside, can e.g. `rm -R /tmp/dietpi-software` at end of script (or best would be exit "trap"), and don't need to take care about possible duplicated file names across different scripts. Especially for `dietpi-software` this is beneficial, since it starts many sub scripts.
- I think it it is clear that the same script should be never executed in parallel. We could implement (e.g. within `G_INIT`) a check if this script already runs an exit in case with meaningful error message. By this we are again more bullet prove against script interferences and securely can remove whole working dir on exit.
-  We can by times roll out some consistent handling of this (script already runs test, tmp working dir, rm on exit "trap") to all scripts, at least I would like that. But here I start with `/tmp/dietpi-software` for now.

MichaIng committed:
Merge branch 'testing' into dietpi-software

MichaIng committed:
v6.12

+ DietPi-Software | Minor alignment

MichaIng committed:
v6.12

+ DietPi-Software | Download_Install: Implement $URL_LIST to allow pre-checking all URLs each software title, before starting any download
+ DietPi-Software | Switch to $FP_WORK_DIR=/tmp/dietpi-software for all file creations and downloads of this script

MichaIng committed:
v6.12

+ DietPi-Software | Fixes + minor wording and coding

MichaIng committed:
v6.12

+ DietPi-Software | Download_Install: Implementation Phase 4

MichaIng commented:
Will merge first, to go on work on updated dietpi-software with some user changes included.

MichaIng committed:
v6.12

+ DietPi-Software | Sickrage: Fix installation failing due to not yet chowned directory
+ DietPi-Software | Remove $FP_WORK_DIR (/tmp/dietpi-software) completely on exit
+ DietPi-Software | Minor: Resolve irrelevant error messages

MichaIng committed:
Merge branch 'testing' into dietpi-software

MichaIng committed:
v6.12

+ DietPi-Software | Remove doubled APT update & upgrade banner
+ DietPi-Software | Download_Install: Use $G_ERROR_HANDLER_EXITCODE_RETURN to catch error code from G_RUN_CMD

MichaIng committed:
v6.12

+ DietPi-Software | Linuxdash: Revert to copy updated files in place instead of directly unzipping to target directory, to prevent possible error with existing directory

MichaIng commented:
@Fourdee 
Made some careful reviews and tested everything I could. Methods are proven to work now, I carefully checked that they were implemented correctly on every updated install. But to be failure safe, testing of the RPi-only installs would be great.
I want to merge the current stage before going on with implementation, to avoid conflicts with the made run user changes.

OpenBazaar install btw works great now, takes max 250M all together now instead of 650M 👍. 

One last thing I would like to have more consistent is how installs are moved/copied/extracted to target location, e.g. on reinstalls. Currently many installs use one of:
- `unzip file.zip -d /target/dir` or `mv dir /target/dir` which both fails, if the dir is already present
- `mv dir/* /target/dir/` which fails, if the folder is present and contains sub dirs
- `cp -R dir/* /target/dir/` which works in every case, as only `cp` is able to merge destination into existing target directory structure

But this will leave old files in place. Some installs remove the whole old install dir because of this, but then potentially configuration files or data is lost.

For few installs we do some config file backup or skip installation, if install dir was already found (did this for ownCloud/Nextcloud e.g.).

Ideally we go through all installs and check which ones have:
- `internal updaters`, to never overwrite existing install directories in those cases, as e.g. downgrading from manually chosen beta branch to lower stable could lead to data/database being unreadable etc. and there is no need to upgrade from our side
- `variable data or configuration files within install dir`, to backup (and recover) those, or merge new (non variable/config) files into old dir, or as well skip overwriting install dir completely
- only completely clean and install fresh, if really just binaries and non variable files are inside the install dir
- and/or give user always the decision to skip, merge or clean and overwrite old installation.

With this PR I left the previous handling mostly the same, just changed some `mv dir/*` to `cp -R dir/*` which at least will not fail on merging sub directories.

MichaIng committed:
v6.12

+ DietPi-Software | RPi Cam: Revert/fix "/etc/raspimjpeg" being the config file itself, not the config folder

MichaIng committed:
v6.12

+ DietPi-Software | Download_Install: Only "tar -C target_directory" needs the target directory to exist, "unzip -d" and "7za -o" will create them automatically

Fourdee commented:
@MichaIng 

Great work on this, looks good 👍 

> Made some careful reviews and tested everything I could. Methods are proven to work now, I carefully checked that they were implemented correctly on every updated install. But to be failure safe, testing of the RPi-only installs would be great.
I want to merge the current stage before going on with implementation, to avoid conflicts with the made run user changes.

Ideally, i'd like to nudge this forward to v6.13 if possible? We urgently need to get v6.12 out due to cron issue and sonarr/radarr etc.

After that, i'll take a look and get this merged, so we can test away. Is that ok?

Fourdee commented:
@MichaIng 

Ah, does this also contain a few fixes for the software titles/installs?

MichaIng commented:
@Fourdee 
Jep, I thought so as well, lets get v6.12 release fast and then have time to finish this PR 👍.

Lets see about included install fixes:

The only important fix was:
- https://github.com/Fourdee/DietPi/pull/1906/commits/64daad3c29b498de11a2af8fa07b69bf219cc36a#diff-d92a6ee04e02fd2a2dc23d5bec3e6a98R5634 As we start the service within configuration step, the folder needs to be chowned here already. Global folder permissions function runs after configuration, thus too late in this case.

Other minor fixes:
- https://github.com/Fourdee/DietPi/pull/1906/commits/2fa796a7480474ddd2a96b0e737961299d35fffb Minor, as this is only Buster support.
- https://github.com/Fourdee/DietPi/pull/1906/commits/de0b0ae56fecbbc09590808242aef7408413a074#diff-d92a6ee04e02fd2a2dc23d5bec3e6a98R5020 (only the certbot additions) Also minor, as this is only relevant in rare cases and certbot can be easily reinstalled without account/cert loose, if it got auto-removed.

Fourdee commented:
@MichaIng 

Ok good stuff 👍 . 

🈯️ I'll apply the major fix now and remove line from this PR

Fourdee committed:
Minor

Applied via https://github.com/Fourdee/DietPi/commit/f7565019d51e831c465cb021b6c5754d75181470

MichaIng committed:
Merge branch 'testing' into dietpi-software

MichaIng commented:
@Fourdee 
Was the closing intended?

MichaIng committed:
Merge branch 'testing' into dietpi-software

Fourdee commented:
@MichaIng 

> Was the closing intended?

Apologies, I'am blaming my 11 month old on this one 😄, he loves keyboards for some reason 😕

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -3270,6 +3277,80 @@ _EOF_
 
 	}
 
+	Download_Install(){
```
Very nice 👍 

Always had the idea to do something similar, but never implemented. Great stuff!

Fourdee commented:
@MichaIng 

Impressive 👍 , looks good, lets merge and test.
